aidial_adapter_bedrock/llm/truncate_prompt.py

```diff
@@ -0,0 +1,106 @@
+from abc import ABC, abstractmethod
+from typing import Callable, List, Optional, Set, TypeVar
+
+from pydantic import BaseModel
+
+from aidial_adapter_bedrock.utils.list import select_by_indices
+
+
+class TruncatePromptError(ABC, BaseModel):
+    @abstractmethod
+    def print(self) -> str:
+        pass
+
+
+class InconsistentLimitsError(TruncatePromptError):
+    user_limit: int
+    model_limit: int
+
+    def print(self) -> str:
+        return (
+            f"Maximum prompt tokens ({self.user_limit}) "
+            f"exceeds the model maximum prompt tokens ({self.model_limit})."
+        )
+
+
+class ModelLimitOverflow(TruncatePromptError):
+    model_limit: int
+    token_count: int
+
+    def print(self) -> str:
+        return (
+            f"Token count of all messages ({self.token_count}) exceeds"
+            f" the model maximum prompt tokens ({self.model_limit})."
+        )
+
+
+class UserLimitOverflow(TruncatePromptError):
+    user_limit: int
+    token_count: int
+
+    def print(self) -> str:
+        return (
+            "Token count of the last message and all system messages "
+            f"({self.token_count}) exceeds the maximum prompt tokens ({self.user_limit})."
+        )
+
+
+T = TypeVar("T")
+Messages = List[T]
+
+
+def truncate_prompt(
+    messages: Messages,
+    tokenize: Callable[[Messages], int],
+    keep_message: Callable[[Messages, int], bool],
+    model_limit: Optional[int],
```

Comment:
The model limit is always None in production code.

Reply:
In the MR it's always None indeed. But when I add model limits to the adapter, it will take a non-trivial value.